library/ssl_tls.c

```diff
@@ -6978,4 +6978,139 @@ int mbedtls_ssl_get_key_exchange_md_tls1_2( mbedtls_ssl_context *ssl,
 
 #endif /* MBEDTLS_SSL_PROTO_TLS1_2 */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+static int ssl_get_handshake_transcript( psa_hash_operation_t *operation,
+                                         unsigned char *dst,
+                                         size_t dst_len,
+                                         size_t *olen )
+{
+    psa_status_t status;
+    psa_hash_operation_t psa_ctx = psa_hash_operation_init();
+
+    status = psa_hash_clone( operation, &psa_ctx );
+    if( status != PSA_SUCCESS )
+    {
+        return( MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED );
+    }
+
+    status = psa_hash_finish( &psa_ctx, dst, dst_len, olen );
+    if( status != PSA_SUCCESS )
+    {
+        return( MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED );
+    }
+
+    return( 0 );
+}
+#else /* MBEDTLS_USE_PSA_CRYPTO */
+typedef union
+{
+#if defined(MBEDTLS_SHA384_C)
+    mbedtls_sha512_context sha512;
+#endif /* MBEDTLS_SHA384_C */
+#if defined(MBEDTLS_SHA256_C)
+    mbedtls_sha256_context sha256;
+#endif /* MBEDTLS_SHA224_C */
+} ssl_hs_ts_fin_hash;
+
+typedef void ( *ssl_hs_ts_init )( void * );
+typedef void ( *ssl_hs_ts_free )( void * );
+typedef void ( *ssl_hs_ts_clone )( void *, void * );
+typedef int ( *ssl_hs_ts_finish )( void *, void * );
+
+typedef struct
+{
+    uint16_t digest_len;
+    ssl_hs_ts_init init;
+    ssl_hs_ts_free free;
+    ssl_hs_ts_clone clone;
+    ssl_hs_ts_finish finish;
+    void * fin_hash_ptr;
+} ssl_hs_ts_operation_t;
+
+static int ssl_get_handshake_transcript( ssl_hs_ts_operation_t *operation,
+                                         unsigned char *dst,
+                                         size_t dst_len,
+                                         size_t *olen )
+{
+    int ret;
+    ssl_hs_ts_fin_hash fin_hash;
+
+    if( dst_len < operation->digest_len )
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+
+    operation->init( &fin_hash );
+    operation->clone( &fin_hash, operation->fin_hash_ptr );
+
+    if( ( ret = operation->finish( &fin_hash, dst ) ) != 0 )
+    {
+        goto exit;
+    }
+
+    *olen = operation->digest_len;
+
+exit:
+
+    operation->free( &fin_hash );
+    return( ret );
+}
+#endif /* !MBEDTLS_USE_PSA_CRYPTO */
+
+int mbedtls_ssl_get_handshake_transcript( mbedtls_ssl_context *ssl,
```

Comment:
The factorization of the code works well for the PSA crypto version but not when using the Mbed TLS APIs. Thus I would suggest to have two versions of the function, one for PSA crypto, one for Mbed TLS APIs
thus something like:
```C
#if defined(MBEDTLS_USE_PSA_CRYPTO)
int mbedtls_ssl_get_handshake_transcript( mbedtls_ssl_context *ssl,
                                          const mbedtls_md_type_t md,
                                          unsigned char *dst,
                                          size_t dst_len,
                                          size_t *olen )
{
...
}
#else /* MBEDTLS_USE_PSA_CRYPTO */
int mbedtls_ssl_get_handshake_transcript( mbedtls_ssl_context *ssl,
                                          const mbedtls_md_type_t md,
                                          unsigned char *dst,
                                          size_t dst_len,
                                          size_t *olen )
{
...
}
#endif /* MBEDTLS_USE_PSA_CRYPTO */               
```
and to basically restore the code that was in the previous version when calling the Mbed TLS APIs.

Reply:
I thought `MBEDTLS_USE_PSA_CRYPTO` was out of scope for the MVP? (That is, that the MVP always uses the legacy APIs, regardless of whether `MBEDTLS_USE_PSA_CRYPTO` is enabled - as do most parts of the existing code anyway.) 

Having two versions of the code certainly doesn't look very minimal to me. So, we should either only use the legacy APIs, or only use the PSA Crypto APIs. I think it was decided some time ago that it would be only the legacy APIs for various reasons. Anyway, the current code for ECDH in the ClientHello is based on the legacy API (and doesn't have a version that uses the PSA Crypto API), so at least for consistency it would make sense to stick to that for other crypto operations as well.

(That's another point of the technical definition of MVP that should be written down.)

Reply:
Good point. Let's remove the PSA crypto code from this PR then. When we merge back this to the prototype we will probably have to do the rework I suggested just above but that's ok with me.